pilot/pkg/networking/core/v1alpha3/tracing.go

```diff
@@ -0,0 +1,480 @@
+				continue
+			}
+			hcm.Tracing = tcfg
+			providerConfigured = true
+			break
+		}
+	}
+
+	if !providerConfigured {
+		log.Warnf("No provider was configured for tracing")
+		hcm.Tracing = &hpb.HttpConnectionManager_Tracing{}
+		// TODO: transition to configuring providers from proxy config here?
+		// something like: configureFromProxyConfig(tracingCfg, opts.proxy.Metadata.ProxyConfig.Tracing)
+	}
+
+	// gracefully fallback to MeshConfig configuration. It will act as an implicit
+	// parent configuration during transition period.
+	configureSampling(hcm.Tracing, tracingCfg.RandomSamplingPercentage, proxyCfg)
+	configureCustomTags(hcm.Tracing, tracingCfg.CustomTags, proxyCfg)
+
+	// if there ia configured max tag length somewhere, fallback to it.
+	if hcm.GetTracing().GetMaxPathTagLength() == nil && proxyCfg.GetTracing().GetMaxPathTagLength() != 0 {
+		hcm.Tracing.MaxPathTagLength = wrapperspb.UInt32(proxyCfg.GetTracing().MaxPathTagLength)
+	}
+}
+
+// TODO: follow-on work to enable bootstrapping of clusters for $(HOST_IP):PORT addresses.
+
+func configureFromProviderConfig(pushCtx *model.PushContext, meta map[string]interface{},
+	providerCfg *meshconfig.MeshConfig_ExtensionProvider) (*hpb.HttpConnectionManager_Tracing, error) {
+	switch provider := providerCfg.Provider.(type) {
+	case *meshconfig.MeshConfig_ExtensionProvider_Zipkin:
+		return buildHCMTracing(pushCtx, providerCfg.Name, provider.Zipkin.Service, provider.Zipkin.Port, provider.Zipkin.MaxTagLength, zipkinConfigGen)
+	case *meshconfig.MeshConfig_ExtensionProvider_Datadog:
+		return buildHCMTracing(pushCtx, providerCfg.Name, provider.Datadog.Service, provider.Datadog.Port, provider.Datadog.MaxTagLength, datadogConfigGen)
+	case *meshconfig.MeshConfig_ExtensionProvider_Lightstep:
+		return buildHCMTracing(pushCtx, providerCfg.Name, provider.Lightstep.Service, provider.Lightstep.Port, provider.Lightstep.MaxTagLength,
+			func(clusterName string) (*anypb.Any, error) {
+				lc := &tracingcfg.LightstepConfig{
+					CollectorCluster: clusterName,
+					AccessTokenFile:  provider.Lightstep.AccessToken,
+				}
+				return anypb.New(lc)
+			})
+
+	case *meshconfig.MeshConfig_ExtensionProvider_Opencensus:
+		return buildHCMTracingOpenCensus(providerCfg.Name, provider.Opencensus.MaxTagLength, func() (*anypb.Any, error) {
+			oc := &tracingcfg.OpenCensusConfig{
+				OcagentAddress:         fmt.Sprintf("%s:%d", provider.Opencensus.Service, provider.Opencensus.Port),
+				OcagentExporterEnabled: true,
+				IncomingTraceContext:   convert(provider.Opencensus.Context),
+				OutgoingTraceContext:   convert(provider.Opencensus.Context),
+			}
+
+			return anypb.New(oc)
+		})
+
+	case *meshconfig.MeshConfig_ExtensionProvider_Stackdriver:
+		return buildHCMTracingOpenCensus(providerCfg.Name, provider.Stackdriver.MaxTagLength, func() (*anypb.Any, error) {
+			proj, ok := meta[platform.GCPProject].(string)
+			if !ok {
```

Comment:
I don't think project number would work, so do we need the fallback here?

Reply:
Do we have anything to indicate that number doesn't work?  i was under the impression that the cloud APIs took either.

Happy to remove however.

Reply:
Just tried in API explorer. It works actually. Thanks!